lib/src/third_party/diff_match_patch/utils.dart

```diff
@@ -0,0 +1,253 @@
+
+  // Walk the text, pulling out a substring for each line.
+  // text.split('\n') would would temporarily double our memory footprint.
+  // Modifying text would create many large strings to garbage collect.
+  while (lineEnd < text.length - 1) {
+    lineEnd = text.indexOf('\n', lineStart);
+
+    if (lineEnd == -1) {
+      lineEnd = text.length - 1;
+    }
+    line = text.substring(lineStart, lineEnd + 1);
+
+    if (lineHash.containsKey(line)) {
+      chars.writeCharCode(lineHash[line]!);
+    } else {
+      if (lineArray.length == maxLines) {
+        // Bail out at 65535 because
+        // final chars1 =  StringBuffer();
+        // chars1.writeCharCode(65536);
+        // chars1.toString().codeUnitAt(0) == 55296;
+        line = text.substring(lineStart);
+        lineEnd = text.length;
+      }
+      lineArray.add(line);
+      lineHash[line] = lineArray.length - 1;
+      chars.writeCharCode(lineArray.length - 1);
+    }
+    lineStart = lineEnd + 1;
+  }
+  return chars.toString();
+}
+
+/// Split two texts into a list of strings. Reduce the texts to a string of
+/// hashes where each Unicode character represents one line.
+///
+/// [text1] is the first string.
+/// [text2] is the second string.
+/// Returns a Map containing the encoded text1, the encoded text2 and
+/// the List of unique strings. The zeroth element of the List of
+/// unique strings is intentionally blank.
+@visibleForTesting
+Map<String, dynamic> diffLinesToChars(String text1, String text2) {
+  final lineArray = <String>[];
+  final lineHash = HashMap<String, int>();
+  // e.g. linearray[4] == 'Hello\n'
+  // e.g. linehash['Hello\n'] == 4
+
+  // '\x00' is a valid character, but various debuggers don't like it.
+  // So we'll insert a junk entry to avoid generating a null character.
+  lineArray.add('');
+
+  // Allocate 2/3rds of the space for text1, the rest for text2.
+  final chars1 = _diffLinesToCharsMunge(text1, lineArray, lineHash, 40000);
+  final chars2 = _diffLinesToCharsMunge(text2, lineArray, lineHash, 65535);
+
+  return {'chars1': chars1, 'chars2': chars2, 'lineArray': lineArray};
+}
+
+/// Compute the Levenshtein distance; the number of inserted, deleted or
+/// substituted characters.
+/// [diffs] is a List of Diff objects.
+/// Returns the number of changes.
+int diffLevenshtein(List<Diff> diffs) {
+  var levenshtein = 0;
+  var insertions = 0;
+  var deletions = 0;
+
+  for (var aDiff in diffs) {
+    switch (aDiff.operation) {
+      case Operation.insert:
+        insertions += aDiff.text.length;
+        break;
+      case Operation.delete:
+        deletions += aDiff.text.length;
+        break;
+      case Operation.equal:
+        // A deletion and an insertion is one substitution.
+        levenshtein += max(insertions, deletions);
+        insertions = 0;
+        deletions = 0;
+        break;
+    }
+  }
+
+  levenshtein += max(insertions, deletions);
+  return levenshtein;
+}
+
+int diffLevenshteinWord(Iterable<Diff> diffs) {
```

Comment:
I don't think this is the actual minimum number of edits to go from a to b. It rather depends on what `Diffs` the diffing representation used to represent it with.

I think we should call this something different or actually compute the levenshtein distance: https://en.wikipedia.org/wiki/Damerau%E2%80%93Levenshtein_distance

Reply:
Should we call this `word-level edit distance` if it makes sense?

Reply:
If it is the "word-level edit distance", then that's a great name... is it that?

(I suspect so, but I could be wrong)